File: collage_sem_2/python_reveiw/inheritance.py
'''
* ilmu baru:
1. Perbedaan Penggunaan super() dan namaClass():
* Intinya penggunaan super() dalam Python adalah cara khusus untuk memanggil metode/objek dari kelas induk, terutama dalam konteks inheritance.
* sementara namaClass()merujuk pada referensi class yang menjadi anak.
* super()fungsi yang akan membuat kelas anak mewarisi semua metode dan properti dari induknya.
* super() itu pemanggilan sebuah class saja tidak bisa mewakili banyak kelas jika dalam kelas yang diwakili masing-masing tidak didefine si super()  
* jika namaClass() itu bisa di define untuk mewakili sebuah kelas, tapi dalam kelas yang diwakili/child class tidak perlu define className() masing-masing. 
2. Sebuah child class bisa di define dengan hanya:
class Student(Person):
  pass
Namun, Lebih baik menggunakan function  __init__() karena function tersebut lebih baik dipakai dari pada keyword pass.
Selain itu,  __init__() mengotomatiskan setiap kali class digunakan untuk membuat objke baru di child class.
 __init__() membuat  si class child tidak lagi mewarisi induknya karena dia mengambil alih warisan  __init__() fuction class induk.
3. Bedakan single inheritance dan multiple inherintance:
ex single inherin:
class Person:
  def __init__(self, fname, lname):
    self.firstname = fname
    self.lastname = lname

  def printname(self):
    print(self.firstname, self.lastname)

class Student(Person):
  def __init__(self, fname, lname, year):
    super().__init__(fname, lname)
    self.graduationyear = year

  def welcome(self):
    print("Welcome", self.firstname, self.lastname, "to the class of", self.graduationyear)

# pemanggilan keseluruhan class single inheritance
x = Student("Mike", "Olsen", 2019)
x.printname() 
x.welcome()


ex: multiple inherin
class Car:
  def __init__(self, brand, model):
    self.brand = brand
    self.model = model

  def move(self):
    print("Drive!")

class Boat:
  def __init__(self, brand, model):
    self.brand = brand
    self.model = model

  def move(self):
    print("Sail!")

class Plane:
  def __init__(self, brand, model):
    self.brand = brand
    self.model = model

  def move(self):
    print("Fly!")

class Output(Car, Boat, Plane):
  def __init__(self, brand, model):
    Car.__init__(self, brand, model)
    Boat.__init__(self, brand, model)
    Plane.__init__(self, brand, model)

  def move_all(self):
    Car.move(self)
    Boat.move(self)
    Plane.move(self)

#Pemanggilan keseluruhan output class yang dicakup oleh class output sebagai mutiple warisan
vehicle = Output("Tesla", "Model S")
vehicle.move_all()  # Ini akan mencetak "Drive!", "Sail!", dan "Fly!"


Jadi, prinsip si warisan/inheritance ini sama baik single maupun multiple. Setiap child class harus serasi struktunya dengan parent class.
Intinya setiap function dari parent itu seperti di duplicate di child dengan perbedaan penambahan objek dan perbedaan nama method.
Konsep paling bagus dalam oop itu pada bagian pemanggilan
'''


# from abc import ABC, abstractmethod


# class Shape(ABC):
#     @abstractmethod
#     def area(self):
#         pass


# class Circle(Shape):
#     def __init__(self, radius):
#         self.radius = radius

#     def area(self):
#         return 3.14 * self.radius * self.radius


# # Creating an object of the Circle class
# circle1 = Circle(5)
# print("Circle Area:", circle1.area())

'''
* Abce adalah module yang disediakan oleh python.
* Kelas abstrak adalah kelas yang tidak dapat diinstansiasi sendiri dan biasanya berfungsi sebagai kerangka dasar untuk kelas lain.
anggap aja kayak template buat digunain/ditambahin input.
* @abstractmethod adalah indikasi bahwa dia adalah class abstract
* Agar tidak menjadi abstract maka class circle harus memiliki objek dan return objeknya sendiri.
'''


# class Car:
#     def __init__(self, brand, model):
#         self.brand = brand
#         self.model = model

#     @property
#     def info(self):
#         return f"{self.brand} {self.model}"

#     @staticmethod
#     def production_year():
#         return 2022


# # Creating an object of the Car class
# car1 = Car("Toyota", "Camry")
# print(car1.info)
# print("Production Year:", Car.production_year())

'''
* @property digunakan untuk membuat metode dalam kelas berperilaku seperti atribut.
*  Ini memungkinkan kita untuk mengakses metode sebagai atribut tanpa perlu memanggilnya sebagai fungsi (tanpa tanda kurung).
* buktinya tidak ada kan di pemanggilan memanggil fungsi info()
* ibarat dia tu membuat kelogisan pemanggilan format argument kelas 
* @staticmethod digunakan untuk mendefinisikan metode yang tidak mengoperasikan instance kelas atau kelas itu sendiri. 
* Jadi dia hanya untuk di return outputnya dan boleh berupa variabel bervalue. intinya dia harus ada value.
* pemanggilan membuat dia tidak bergantung dengan objek class manapun

'''


'''
TO-do:
2. mencari tahu penjelasan __str__() dan __add__()
'''


class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __str__(self):
        return f"Point({self.x}, {self.y})"

    def __add__(self, other):
        return Point(self.x + other.x, self.y + other.y)


# Creating an object of the Point class
point1 = Point(1, 2)
print(point1)
point2 = Point(3, 4)
print(point2)

# Using the + operator
point3 = point1 + point2
print(point3)

File: collage_sem_2/java_review/inheritance.java
/**
 * Secara garis besar ada dua tipe, yaitu subclass(child) dan
 * supperclass(parent)
 * Keyword yang digunakan adalah expand(kelas parent)
 * Protected modifer digunakan untuk membuat sebuah objek yang didefine agar
 * privat sehingga tidak dapat diakses oleh kelas lain kecuali inheritance.
 * final modifer juga ada. gunanya agar suatu class dalam oop itu tidak
 * menggunakan/menyertakan suatu class. Kayak diskip aja tu class.
 */

/**
 * Penggunaan this dan apa itu.
 * 1. referensi variabel yang milik class parent.
 * 3. referensi variabel anatar constructor method.
 */

/*
 * // To-do: tambah ilmu dari ppt collage
 * class parent {//kenapa code class parent ini bisa satu file dengan main
 * classnya? karena dia bukan public class melainkan default sehingga bisa.
 * public int x = 5; //constructor parent
 * }
 * 
 * class Child extends parent {
 * public int x = 10; //constructor child yang mana harus di define ulang
 * variabelnya karena dia bersifat tidak mewarisi induknya. Kalo lupa tenang aja
 * otomatis dibantu compiler java.
 * 
 * public void Info(int x) {
 * System.out.println("Nilai x sebagai parameter = " + x);// merujuk pada
 * variabel x terdekat, yaitu parameter info() karena logikanya argumen x pada
 * info(x) yang dilakukan pemanggilan di kelas main ya merujuk ke dia.
 * 
 * System.out.println("Data member x di class Child = " + this.x);//merujuk pada
 * variabel class child karena this menargetkan variabel yang hanya ada di kelas
 * itu sendiri. sehingga ketika
 * //terjadi situasi di mana variabel dengan nama yang sama ada di kelas induk
 * dan kelas anak, itu dapat membantu membedakan rujukan variabel lokal dan
 * variabel yang diwarisi(dengan cara super())
 * 
 * System.out.println("Data member x di class Parent = " + super.x);//merujuk
 * pada variabel class parent karena sebagai sifat pewarisan oleh induk ke anak.
 * Ini melanjutkan kejadian sebelumnya agar dapat membedakan variabel yang sama
 * namanya di kelas parent dna child.
 * }
 * }
 * 
 * public class inheritance {// main class dan harus sama namanya dengan file.
 * public static void main(String args[]) {
 * Child tes = new Child();
 * tes.Info(20);
 * }
 * }
 */
// Terapan kedua

// INFO: ini menggunakan Parent class public sehingga file terpisah. karena
// aturan dari java.

class Baby extends Parent {
    public Baby(String parentName) {
        // super(parentName); // code cara Memanggil konstruktor Parent karena
        // constructor itu kayak class.
    }

    public void Cry() {
        System.out.println("Owek owek");
    }
}

public class inheritance {
    public static void main(String[] args) {
        Baby bayi = new Baby("NamaParent"); // Sekarang ini valid
        bayi.Cry();
    }
}